mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoSafeUpdateProtocol.java

```diff
@@ -0,0 +1,349 @@
+                index=0;
+                for(Object id:idsInBatch) {
+                    if(!results.containsKey(index)) { // No other errors for this id
+                        if(!successfulIds.contains(id)) {
+                            // concurrency errors for this id
+                            results.put(index,Error.get("update",MongoCrudConstants.ERR_CONCURRENT_UPDATE,id.toString()));
+                        }
+                    }
+                    index++;
+                }
+            } finally {
+                if(cursor!=null)
+                    cursor.close();
+            }
+        }
+    }
+
+    /**
+     * Writes a replace query, which is:
+     * <pre>
+     *  {_id:<docId>, "@mongoHidden.docver.0":ObjectId(<originalDocver>)}
+     * </pre>
+     */
+    private DBObject writeReplaceQuery(DBObject doc) {
+        DBObject hidden=getHidden(doc,false);
+        ObjectId top=null;
+        if(hidden!=null) {
+            List<ObjectId> list=(List<ObjectId>)hidden.get(DOCVER);
+            if(list!=null&&!list.isEmpty())
+                top=list.get(0);
+        }
+        BasicDBObject query=new BasicDBObject("_id",doc.get("_id"));
+        query.append(DOCVER_FLD0,top);
+        return query;
+    }
+    
+
+    /**
+     * Returns the @mongoHidden at the root level of the document. Adds one if necessary.
+     */
+    public static DBObject getHidden(DBObject doc,boolean addIfNotFound) {
+        DBObject hidden=(DBObject)doc.get(Translator.HIDDEN_SUB_PATH.toString());
+        if(hidden==null&&addIfNotFound) {
+            doc.put(Translator.HIDDEN_SUB_PATH.toString(),hidden=new BasicDBObject());            
+        }
+        return hidden;
+    }
+
+    public static void overwriteDocVer(DBObject doc,ObjectId docver) {
+        getHidden(doc,true).removeField(DOCVER);
+        setDocVer(doc,docver);
+    }
+    
+    public static void setDocVer(DBObject doc,ObjectId docver) {
+        DBObject hidden=getHidden(doc,true);
+        List<ObjectId> list=(List<ObjectId>)hidden.get(DOCVER);
+        if(list==null) {
+            list=new ArrayList<ObjectId>();
+        }
+        list.add(0,docver);
+        hidden.put(DOCVER,list);
+    }
+
+    /**
+     * Removes old docvers from a document
+     */
+    private void cleanupOldDocVer(DBObject doc) {
+        DBObject hidden=getHidden(doc,false);
+        if(hidden!=null) {
+            List<ObjectId> list=(List<ObjectId>)hidden.get(DOCVER);
+            if(list!=null) {
+                List<ObjectId> copy=new ArrayList<>(list.size());
+                long now=docVer.getDate().getTime();
+                for(ObjectId id:list) {
+                    if(!id.equals(docVer)) {
+                        Date d=id.getDate();
+                        if(now-d.getTime()<TOO_OLD) {
+                            copy.add(id);
+                        }
+                    } else {
+                        copy.add(id);
+                    }
+                }
```

Comment:
Yes, timeout is imperfect. The timeout is between the time A commits, and
reads it back, so if A cannot read back what it wrote within a minute, and
some other process comes and updates it, A will think it is a concurrent
update. With the existing mongo apis, the only way to deal with this is to
increase the timeout.

I think 1 minute timeout is large enough to make this scenario rare enough,
but we can increase it.

On Fri, Aug 26, 2016 at 1:46 PM, Alec Henninger <notifications@github.com>
wrote:

> In mongo/src/main/java/com/redhat/lightblue/mongo/crud/
> MongoSafeUpdateProtocol.java
> <https://github.com/lightblue-platform/lightblue-mongo/pull/303#discussion_r76477103>
> :
>
> > +        DBObject hidden=getHidden(doc,false);
> > +        if(hidden!=null) {
> > +            List<ObjectId> list=(List<ObjectId>)hidden.get(DOCVER);
> > +            if(list!=null) {
> > +                List<ObjectId> copy=new ArrayList<>(list.size());
> > +                long now=docVer.getDate().getTime();
> > +                for(ObjectId id:list) {
> > +                    if(!id.equals(docVer)) {
> > +                        Date d=id.getDate();
> > +                        if(now-d.getTime()<TOO_OLD) {
> > +                            copy.add(id);
> > +                        }
> > +                    } else {
> > +                        copy.add(id);
> > +                    }
> > +                }
>
> Hrm... concurrency is always a bit mind bending :).
>
> The scenario that could be a problem (but perhaps it is too rare depending
> on timeout?) I think is this:
>
> Given timeout is 1:00 (1 minute), and there are [at least] three
> overlapping updates throughout this timeframe.
>
>    1. Two updates, A and B find their docs at ~0:00
>    2. A and B start commits at 0:01 with docvers from ~0:00
>    3. They finish. B is successful, returns. A misses some updates, will
>    check for concurrent mods.
>    4. Update C starts at 1:00, removes some docvers from update A which
>    are too old, and commits.
>    5. A finds concurrent mods at 1:01 (big delay from step 2... would
>    this happen?), gets less back than expected because of update C removing
>    old doc versions. It knows it updated N documents, but finds >N concurrent
>    mods.
>
> What do you think? Maybe to some extent we can't be perfect, and if
> timeout is long enough, it would be crazy to actually happen. So maybe
> imperfect is okay. (Assuming I don't have something wrong). Even my
> suggestion (keeping some old ones if array size not too big) is still
> imperfect (just imagine there are many successful updates in between 2 and
> 4 ... array size could be met eventually).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lightblue-platform/lightblue-mongo/pull/303/files/6e592caaaf4c79fd821914fdb9640896fa145ab0#r76477103>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADgDDSqjLbW7J2H3O2yzcX4Jlz0tCa-4ks5qj0KLgaJpZM4JtbtZ>
> .
>